src/utilities.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utilities used by the charm."""
+
+import logging
+import subprocess  # nosec B404
+import time
+from typing import Callable, Optional, Sequence, Type, TypeVar
+
+logger = logging.getLogger(__name__)
+
+# Return type of decorated function
+R = TypeVar("R")
+
+
+def retry(
+    exception: Type[Exception] = Exception,
+    tries: int = 1,
+    delay: float = 0,
+    max_delay: Optional[float] = None,
+    backoff: float = 1,
+    logger: logging.Logger = logger,
+) -> Callable[[Callable[..., R]], Callable[..., R]]:
+    """Parameterize the decorator for adding retry to functions.
+
+    Args:
+        exception: Exception type to be retried. Defaults to Exception.
+        tries: Number of attempts at retry. Defaults to 1.
+        delay: Time in seconds to wait between retry. Defaults to 0.
+        max_delay: Max time in seconds to wait between retry. Defaults to None.
+        backoff: Factor to increase the delay by each retry. Defaults to 1.
+        logger: Logger for logging. Defaults to module logger.
+
+    Returns:
+        The function decorator for retry.
+
+    TODO:
+    For Python 3.10+, PEP 612 allows to specify the parameter of callable using `ParamSpec`. Do
+    this when focal (Python 3.8) is no longer supported.
+    """
+
+    def retry_decorator(
+        fn: Callable[..., R],
+    ) -> Callable[..., R]:
+        """Decorate function with retry.
+
+        Args:
+            fn (Callable[..., R]): The function to decorate.
+
+        Returns:
+            Callable[..., R]: The resulting function with retry added.
+        """
+
+        def fn_with_retry(*args, **kwargs) -> R:
+            """Function with retry"""
+            remain_tries, current_delay = tries, delay
+
+            while True:
```

Comment:
what about `for _ in range(tries)`? That way we avoid accidental infinite loops?